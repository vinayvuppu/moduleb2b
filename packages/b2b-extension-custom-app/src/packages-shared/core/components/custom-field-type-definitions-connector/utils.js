import { transformLocalizedFieldToString } from '@commercetools-local/utils/graphql';
import omit from 'lodash.omit';

export const restDocToForm = customFieldsFromRestResponse => ({
  fields: customFieldsFromRestResponse.fields,
  type: {
    id: customFieldsFromRestResponse.type.id,
    ...customFieldsFromRestResponse.type.obj,
  },
});

export const graphQlDocToForm = typeDefinition => ({
  fields: {},
  type: {
    id: typeDefinition.id,
    key: typeDefinition.key,
    name: transformLocalizedFieldToString(typeDefinition.nameAllLocales),
    fieldDefinitions: typeDefinition.fieldDefinitions.map(fieldDefinition => ({
      name: fieldDefinition.name,
      required: fieldDefinition.required,
      type: do {
        if (fieldDefinition.type?.elementType?.name === 'LocalizedEnum')
          ({
            ...fieldDefinition.type,
            elementType: {
              ...fieldDefinition.type.elementType,
              values: fieldDefinition.type.elementType.values.map(value => ({
                key: value.key,
                label: transformLocalizedFieldToString(value.labelAllLocales),
              })),
            },
          });
        else if (fieldDefinition.type?.name === 'LocalizedEnum')
          ({
            ...fieldDefinition.type,
            values: fieldDefinition.type.values.map(value => ({
              key: value.key,
              label: transformLocalizedFieldToString(value.labelAllLocales),
            })),
          });
        else fieldDefinition.type;
      },
      label: transformLocalizedFieldToString(fieldDefinition.labelAllLocales),
    })),
  },
});

export const formToRestDoc = formCustomFields => ({
  type: {
    id: formCustomFields.type.id,
    typeId: 'type',
    obj: {
      ...formCustomFields.type,
    },
  },
  fields: formCustomFields.fields,
});

export const formToGraphQlDoc = formCustomFields => ({
  type: {
    id: formCustomFields.type.id,
    typeId: 'type',
    obj: {
      ...omit(formCustomFields.type, ['id']),
    },
  },
  fields: formCustomFields.fields || {},
});

export const createEmptyCustomFields = () => ({
  fields: {},
  type: { fieldDefinitions: [] },
});
