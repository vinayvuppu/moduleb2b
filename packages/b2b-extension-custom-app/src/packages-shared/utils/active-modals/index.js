export { default } from './active-modals';
