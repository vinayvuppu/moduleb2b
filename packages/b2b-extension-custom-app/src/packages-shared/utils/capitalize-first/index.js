export { default } from './capitalize-first';
