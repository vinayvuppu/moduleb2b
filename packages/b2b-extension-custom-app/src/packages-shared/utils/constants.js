export const PRODUCT_STATUSES = {
  PUBLISHED: 'published',
  UNPUBLISHED: 'unpublished',
  MODIFIED: 'modified',
};

export const FILTER_TYPES = {
  lessThan: 'lessThan',
  moreThan: 'moreThan',
  equalTo: 'equalTo',
  range: 'range',
  missing: 'missing',
  missingIn: 'missingIn',
  in: 'in',
};

export const FIELD_TYPES = {
  Money: 'Money',
  LocalizedString: 'LocalizedString',
  Time: 'Time',
  DateTime: 'DateTime',
  Boolean: 'Boolean',
  String: 'String',
  Enum: 'Enum',
  LocalizedEnum: 'LocalizedEnum',
  Number: 'Number',
  Date: 'Date',
  Reference: 'Reference',
  Set: 'Set',
};

export const PRECISION_TYPES = {
  highPrecision: 'highPrecision',
  centPrecision: 'centPrecision',
};

export const DefaultPageSizes = [20, 50, 100];
export const DefaultPageSize = DefaultPageSizes[0];
