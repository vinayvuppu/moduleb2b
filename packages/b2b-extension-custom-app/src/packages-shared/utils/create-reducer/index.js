export { default } from './create-reducer';
