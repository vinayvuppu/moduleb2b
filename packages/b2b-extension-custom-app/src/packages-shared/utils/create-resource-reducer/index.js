export { default } from './create-resource-reducer';
