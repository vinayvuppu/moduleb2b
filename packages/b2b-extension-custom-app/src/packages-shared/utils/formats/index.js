export { formatDateTime, formatDateRangeValue } from './date';
export { formatMoney, formatMoneyValue, formatMoneyRangeValue } from './money';
export { default as formatProductAttribute } from './product-attribute';
export { default as formatCustomField } from './custom-field';
