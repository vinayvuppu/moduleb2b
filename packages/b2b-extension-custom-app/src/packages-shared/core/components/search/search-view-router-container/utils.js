import { getIn } from 'formik';
import omit from 'lodash.omit';
import { encode as encodeQueryString, decode as decodeQueryString } from 'qss';
import { stringToBase64, base64ToString } from '../../../../utils/encodings';

/**
 * NOTE
 *   The `version` comes first and can be overwriten to bust
 *   existing search queries in case of conflicting migrations.
 */
export const VERSION_KEY = '__version';
export const matchesVersion = (searchQuery, version) =>
  Boolean(searchQuery[VERSION_KEY]) && searchQuery[VERSION_KEY] === version;
export const augmentWithVersion = (searchQuery, version) => ({
  [VERSION_KEY]: version,
  ...searchQuery,
});
export const omitVersion = searchQuery => omit(searchQuery, [VERSION_KEY]);

export const hasQuery = location =>
  getIn(location, 'query.query', '').length > 0;
export const hasStorage = storageSlice =>
  window.localStorage.getItem(storageSlice) !== null;

export const encode = searchOptions =>
  stringToBase64(JSON.stringify(searchOptions));
export const decode = searchOptions =>
  searchOptions && JSON.parse(base64ToString(searchOptions));

export const getFromStorage = storageSlice =>
  decode(window.localStorage.getItem(storageSlice));
export const putToStorage = (storageSlice, searchOptions) =>
  window.localStorage.setItem(storageSlice, encode(searchOptions));
export const removeFromStorage = storageSlice =>
  window.localStorage.removeItem(storageSlice);

export const encodeToUrl = (nextSearchOptions, previousSearch) => {
  const previousQueryString = decodeQueryString(previousSearch);
  /**
   * NOTE:
   *    `track` needs to be omitted from the `searchOptions`. It is
   *    used in action creators to track what filters are selected
   *    but should not end up in the URL.
   */
  const nextSearchOptionsAsQueryString = encode(
    omit(nextSearchOptions, 'track')
  );
  const nextQueryString = {
    ...previousQueryString,
    ...{
      query: nextSearchOptionsAsQueryString,
    },
  };

  return encodeQueryString(nextQueryString);
};
