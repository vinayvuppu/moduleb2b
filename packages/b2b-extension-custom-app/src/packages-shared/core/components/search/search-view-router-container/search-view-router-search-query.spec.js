import React from 'react';
import { shallow } from 'enzyme';
import { reportErrorToSentry } from '@commercetools-frontend/sentry';
import EnhancedSearchViewRouterSearchQuery, {
  SearchViewRouterSearchQuery,
  withSearchViewRouterSearchQuery,
} from './search-view-router-search-query';
import { VERSION_KEY, omitVersion, encode } from './utils';

jest.mock('@commercetools-frontend/sentry');

const TestComponent = props => <div {...props}>TestComponent</div>;
TestComponent.displayName = 'TestComponent';

const version = 1;
const searchQuery = { foo: 'bar', [VERSION_KEY]: version };
const createTestProps = props => ({
  children: jest.fn(() => <TestComponent />),
  initialSearchQuery: { foo: 'initial-search-query' },
  version,

  // withRouter
  location: {
    search: 'foo-search',
    pathname: 'foo-pathname',
    query: {
      query: encode(searchQuery),
    },
  },
  history: {
    replace: jest.fn(),
    createHref: jest.fn(),
  },

  ...props,
});

describe('rendering', () => {
  let props;
  let wrapper;

  beforeEach(() => {
    props = createTestProps();
    wrapper = shallow(<SearchViewRouterSearchQuery {...props} />);
  });

  it('should invoke `children`', () => {
    expect(props.children).toHaveBeenCalled();
  });

  describe('`searchQuery`', () => {
    it('should pass `searchQuery` in `searchQuery`', () => {
      expect(props.children).toHaveBeenCalledWith(
        expect.objectContaining({
          searchQuery: omitVersion(searchQuery),
        })
      );
    });

    it('should pass `setSearchQuery` in `searchQuery`', () => {
      expect(props.children).toHaveBeenCalledWith(
        expect.objectContaining({
          setSearchQuery: wrapper.instance().setSearchQuery,
        })
      );
    });

    it('should pass `getSearchQuery` in `searchQuery`', () => {
      expect(props.children).toHaveBeenCalledWith(
        expect.objectContaining({
          getSearchQuery: wrapper.instance().getSearchQuery,
        })
      );
    });
  });

  describe('interacting', () => {
    beforeEach(() => {
      window.localStorage.setItem.mockClear();
      window.localStorage.getItem.mockClear();
      window.localStorage.removeItem.mockClear();
    });
    describe('`getSearchQuery`', () => {
      describe('without `location.search`', () => {
        beforeEach(() => {
          wrapper = shallow(
            <SearchViewRouterSearchQuery
              {...createTestProps({
                query: { query: {} },
              })}
            />
          );
        });

        it('should match snapshot', () => {
          expect(wrapper.instance().getSearchQuery()).toMatchSnapshot();
        });
      });

      describe('with `location.query`', () => {
        beforeEach(() => {
          wrapper = shallow(
            <SearchViewRouterSearchQuery {...createTestProps({})} />
          );
        });

        describe('when `query` contains valid JSON', () => {
          let result;
          beforeEach(() => {
            result = wrapper.instance().getSearchQuery();
          });

          it('should match snapshot', () => {
            expect(wrapper.instance().getSearchQuery()).toMatchSnapshot();
          });

          it('should return the `searchQuery` from the `query`', () => {
            expect(result).toEqual({ foo: 'bar' });
          });

          describe('when not all properties are within `result`', () => {
            beforeEach(() => {
              wrapper = shallow(
                <SearchViewRouterSearchQuery
                  {...createTestProps({
                    initialSearchQuery: { bar: 'bar' },
                  })}
                />
              );

              result = wrapper.instance().getSearchQuery();
            });

            it('should default those with `initialSearchQuery`', () => {
              expect(result).toHaveProperty('foo');
              expect(result).toHaveProperty('bar');
            });
          });
        });

        describe('when `query` contains invalid JSON', () => {
          let result;
          beforeEach(() => {
            wrapper = shallow(
              <SearchViewRouterSearchQuery
                {...createTestProps({
                  location: {
                    query: {
                      query: 'asd222--##',
                    },
                  },
                })}
              />
            );

            result = wrapper.instance().getSearchQuery();
          });

          it('should match snapshot', () => {
            expect(wrapper.instance().getSearchQuery()).toMatchSnapshot();
          });

          it('should return the `initialSearchQuery` as the `searchQuery`', () => {
            expect(result).toEqual(props.initialSearchQuery);
          });

          it('should invoke `reportErrorToSentry`', () => {
            expect(reportErrorToSentry).toHaveBeenCalled();
          });
        });
      });

      describe('when `query` does not match version', () => {
        let result;
        beforeEach(() => {
          wrapper = shallow(
            <SearchViewRouterSearchQuery
              {...createTestProps({
                version: 3,
                location: {
                  query: {
                    version: 1,
                    query: 'asd222--##',
                  },
                },
              })}
            />
          );

          result = wrapper.instance().getSearchQuery();
        });

        it('should match snapshot', () => {
          expect(result).toMatchSnapshot();
        });

        it('should default those with `initialSearchQuery`', () => {
          expect(result).toEqual(props.initialSearchQuery);
        });
      });
    });

    describe('`setSearchQuery`', () => {
      describe('with `storageSlice`', () => {
        beforeEach(() => {
          props = createTestProps({
            storageSlice: 'foo-slice',
          });
          wrapper = shallow(<SearchViewRouterSearchQuery {...props} />);

          wrapper.instance().setSearchQuery(searchQuery);
        });

        it('should invoke `put` on `storage`', () => {
          expect(window.localStorage.setItem).toHaveBeenCalled();
        });
      });

      describe('without `storageSlice`', () => {
        beforeEach(() => {
          props = createTestProps({});
          wrapper = shallow(<SearchViewRouterSearchQuery {...props} />);
          wrapper.instance().setSearchQuery(searchQuery);
        });

        it('should not invoke `put` on `storage`', () => {
          expect(window.localStorage.setItem).not.toHaveBeenCalled();
        });

        it('should `replace` on `history`', () => {
          expect(props.history.replace).toHaveBeenCalled();
        });

        it('should `createHref` on `history`', () => {
          expect(props.history.createHref).toHaveBeenCalled();
        });
      });
    });
  });
});

describe('withSearchViewRouterSearchQuery', () => {
  const createArgs = custom => ({
    version: 3,
    storageSlice: jest.fn(() => 'testStorageSlice'),
    ...custom,
  });
  const createApi = props => ({
    searchQuery,
    setSearchQuery: jest.fn(),
    getSearchQuery: jest.fn(),

    ...props,
  });

  describe('without `propName`', () => {
    let wrapper;
    let args;
    let EnhancedComponent;

    beforeEach(() => {
      args = createArgs();
      EnhancedComponent = withSearchViewRouterSearchQuery(args)(TestComponent);
      wrapper = shallow(<EnhancedComponent />);
    });

    it('should match snapshot', () => {
      expect(wrapper).toMatchSnapshot();
    });

    it('should invoke `storageSlice`', () => {
      expect(args.storageSlice).toHaveBeenCalled();
    });

    describe('<SearchViewRouterSearchQuery>', () => {
      let enhancedComponentWrapper;
      let apiMock;

      beforeEach(() => {
        apiMock = createApi();

        enhancedComponentWrapper = wrapper
          .find(EnhancedSearchViewRouterSearchQuery)
          .renderProp('children')(apiMock);
      });

      it('should pass a the `value` of the `searchQuery` prop to the enhanced component', () => {
        expect(enhancedComponentWrapper.prop('searchQuery').value).toEqual(
          searchQuery
        );
      });

      it('should pass a the `get` for the `searchQuery` prop to the enhanced component', () => {
        expect(enhancedComponentWrapper.prop('searchQuery').get).toEqual(
          apiMock.getSearchQuery
        );
      });

      it('should pass a the `set` for the `searchQuery` prop to the enhanced component', () => {
        expect(enhancedComponentWrapper.prop('searchQuery').set).toEqual(
          apiMock.setSearchQuery
        );
      });

      it('should pass a the `version` for the `searchQuery` to the <SearchViewRouterSearchQuery>', () => {
        expect(
          wrapper.find(EnhancedSearchViewRouterSearchQuery).prop('version')
        ).toEqual(args.version);
      });
    });
  });

  describe('with `propName`', () => {
    let wrapper;
    let args;
    let EnhancedComponent;

    beforeEach(() => {
      args = createArgs({
        propName: 'anotherSearchQuery',
        version: 2,
      });
      EnhancedComponent = withSearchViewRouterSearchQuery(args)(TestComponent);
      wrapper = shallow(<EnhancedComponent />);
    });

    it('should match snapshot', () => {
      expect(wrapper).toMatchSnapshot();
    });

    it('should invoke `storageSlice`', () => {
      expect(args.storageSlice).toHaveBeenCalled();
    });

    describe('<SearchViewRouterSearchQuery>', () => {
      let enhancedComponentWrapper;
      let apiMock;

      beforeEach(() => {
        apiMock = createApi();

        enhancedComponentWrapper = wrapper
          .find(EnhancedSearchViewRouterSearchQuery)
          .renderProp('children')(apiMock);
      });

      it('should pass a the `value` of the `searchQuery` prop to the enhanced component', () => {
        expect(
          enhancedComponentWrapper.prop('anotherSearchQuery').value
        ).toEqual(searchQuery);
      });

      it('should pass a the `get` for the `searchQuery` prop to the enhanced component', () => {
        expect(enhancedComponentWrapper.prop('anotherSearchQuery').get).toEqual(
          apiMock.getSearchQuery
        );
      });

      it('should pass a the `set` for the `searchQuery` prop to the enhanced component', () => {
        expect(enhancedComponentWrapper.prop('anotherSearchQuery').set).toEqual(
          apiMock.setSearchQuery
        );
      });

      it('should pass a the `version` for the `searchQuery` to the <SearchViewRouterSearchQuery>', () => {
        expect(
          wrapper.find(EnhancedSearchViewRouterSearchQuery).prop('version')
        ).toEqual(args.version);
      });
    });
  });
});
