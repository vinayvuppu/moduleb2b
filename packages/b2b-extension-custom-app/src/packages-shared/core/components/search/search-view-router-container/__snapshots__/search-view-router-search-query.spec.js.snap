// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`rendering interacting \`getSearchQuery\` when \`query\` does not match version should match snapshot 1`] = `
Object {
  "foo": "initial-search-query",
}
`;

exports[`rendering interacting \`getSearchQuery\` with \`location.query\` when \`query\` contains invalid JSON should match snapshot 1`] = `
Object {
  "foo": "initial-search-query",
}
`;

exports[`rendering interacting \`getSearchQuery\` with \`location.query\` when \`query\` contains valid JSON should match snapshot 1`] = `
Object {
  "foo": "bar",
}
`;

exports[`rendering interacting \`getSearchQuery\` without \`location.search\` should match snapshot 1`] = `
Object {
  "foo": "bar",
}
`;

exports[`withSearchViewRouterSearchQuery with \`propName\` should match snapshot 1`] = `
<withRouter(SearchViewRouterSearchQuery)
  storageSlice="testStorageSlice"
  version={2}
>
  <Component />
</withRouter(SearchViewRouterSearchQuery)>
`;

exports[`withSearchViewRouterSearchQuery without \`propName\` should match snapshot 1`] = `
<withRouter(SearchViewRouterSearchQuery)
  storageSlice="testStorageSlice"
  version={3}
>
  <Component />
</withRouter(SearchViewRouterSearchQuery)>
`;
